// ui/thread.js
var CommentActions = require('./comment-actions')
  , React = require('react')
  ;

var Comment = exports.Comment = React.createClass({
  render: function () {
    var comment = this.props.comment
      , user = this.props.users[this.props.comment.user]
      ;
    var date = new Date(comment.createdAt);
    var userId =  user ? user.id : '54111be98ca49e6f95b202c2';
    var userName = user ? user.name : 'Anonymous';
    var actions = CommentActions({
      parent: comment.id
    });
    return React.DOM.div({className: 'comment', id: 'comment-' + comment.id}
      , React.DOM.div({className: 'comment-header'}
        , React.DOM.a({className: 'user-name', href: '/users/' + userId }, userName)
        , React.DOM.a({className: 'comment-anchor-link', href: '#comment-' + comment.id}
          , React.DOM.span({className: 'comment-timestamp'}
              , React.DOM.time({dateTime: date.toJSON(), title: date.toString()}, date.toDateString())
            )
          )
        )
      , React.DOM.div({className: 'comment-body'}, comment.body)
      , React.DOM.div({className: 'comment-footer'}
        , React.DOM.div({className: 'comment-actions'}, actions)
        )
      , Thread({
        comments: comment.thread || []
        , depth: this.props.depth + 1
        , users: this.props.users
      })
    );
  }
});

var Thread = exports.Thread = React.createClass({
  render: function () {
    var comments = {}
    , depth = this.props.depth
    , users = this.props.users
    ;
    this.props.comments.forEach(function (comment) {
      comments[comment.id] = Comment({
        comment: comment
        , depth: depth
        , users: users
      });
    });

    return React.DOM.div({ className: 'thread thread-depth-' + this.props.depth }, comments);
  }
});
